contrib/fparser/fparser_ad.cc

```diff
@@ -386,15 +387,85 @@ template<typename Value_t>
 int FunctionParserADBase<Value_t>::AutoDiff(const std::string& var_name)
 {
   this->CopyOnWrite();
+  const bool cached = mADFlags & ADCacheDerivatives;
 
   try
   {
-    return ad->AutoDiff(LookUpVarOpcode(var_name), this->mData);
+    unsigned int var_number = LookUpVarOpcode(var_name);
+
+    // should and cen we load a cached derivative?
+    std::string cache_file;
+    const std::string jitdir = ".jitdir";
+    if (cached)
+    {
+      // generate a sha1 hash of the Value type size, byte code, and immediate list
+      SHA1 *sha1 = new SHA1();
+      char result[41]; // 40 sha1 chars plus null
+      size_t value_t_size = sizeof(Value_t);
+      sha1->addBytes((char*) &value_t_size, sizeof(value_t_size));
+      sha1->addBytes((char*) &this->mData->mByteCode[0], this->mData->mByteCode.size() * sizeof(unsigned));
+      if (!this->mData->mImmed.empty())
+        sha1->addBytes((char*) &this->mData->mImmed[0], this->mData->mImmed.size() * sizeof(Value_t));
```

Comment:
Yeah, I'should probably change this. I was sticking to the (ugly) FParser code conventions. But why should I?